src/crimson/os/seastore/nvmedevice/nvmedevice.cc

```diff
@@ -0,0 +1,227 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <sys/mman.h>
+#include <string.h>
+
+#include <fcntl.h>
+
+#include "crimson/common/log.h"
+
+#include "include/buffer.h"
+#include "crimson/os/seastore/nvmedevice/nvmedevice.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+static constexpr uint32_t MAX_EVENTS = 1024;
+
+namespace crimson::os::seastore::nvme_device {
+
+/* background io poller for multi-stream write */
+void poll_completion(std::vector<::io_context_t>* ctxs, bool* exit) {
+  while (*exit == false) {
+    for (auto& ctx : *ctxs) {
+      io_event events[MAX_EVENTS];
+
+      /*
+       * At least a completion should be returned. Otherwise, thread is blocked
+       * until it is possible
+       */
+      int num_events = io_getevents(ctx, 1, MAX_EVENTS, events, NULL);
+
+      for (int i = 0; i < num_events; i++) {
+        io_context_t* io_context = (io_context_t*)events[i].obj;
+        io_context->done = true;
+      }
+    }
+  }
+}
+
+open_ertr::future<>
+NormalNBD::open(const std::string &in_path, seastar::open_flags mode) {
+  /* Open with posix fd for pass generic NVMe commands */
+  fd = seastar::file_desc::open(in_path, (int)mode);
+  identify_controller_data_t controller_data = {0, };
+  return identify_controller(controller_data).safe_then(
+      [this, controller_data, in_path, mode]() {
+      protocol_version = controller_data.version;
+      logger().debug("nvme protocol {}.{} {}",
+          (uint32_t)protocol_version.major_ver,
+          (uint32_t)protocol_version.minor_ver,
+          (uint32_t)protocol_version.tertiary_ver);
+
+      /*
+       * Multi Stream Write
+       *
+       * When NVMe SSD supports multi stream functionality, it marks oacs bit of
+       * identify_controller_data structure (from NVMe Specification 1.4).
+       * If oacs field is true, NormalNBD class opens device file multiple times
+       * with different stream IDs. When user calls write() with stream argument,
+       * NormalNBD finds pre-opened FD with stream ID and submit write IO to the
+       * found FD.
+       */
+      support_multistream = controller_data.oacs.support_directives;
+      if (support_multistream) {
+        write_life_max = 6;
+      }
+
+      open_for_io(in_path, mode);
+
+      /* PWG and PWA are supported from NVMe 1.4 */
+      if (protocol_version.major_ver >= 1 && protocol_version.minor_ver >= 4) {
```

Comment:
This check will fail for a future 2.0 version...

Reply:
Will fix this.